membership-attribute-service/app/controllers/AccountController.scala

```diff
@@ -58,13 +59,75 @@ object AccountHelpers {
 
 }
 
+case class CancellationEffectiveDate(cancellationEffectiveDate: String)
+object CancellationEffectiveDate {
+  implicit val cancellationEffectiveDateFormat = Json.format[CancellationEffectiveDate]
+}
+
 class AccountController(commonActions: CommonActions, override val controllerComponents: ControllerComponents) extends BaseController with LazyLogging {
   import AccountHelpers._
   import commonActions._
   implicit val executionContext: ExecutionContext = controllerComponents.executionContext
 
-  def cancelSubscription[P <: SubscriptionPlan.AnyPlan : SubPlanReads](subscriptionName: memsub.Subscription.Name) =
-    AuthAndBackendViaAuthLibAction.async { implicit request =>
+  /**
+   * fetched with /v1/subscription/{key}?charge-detail=current-segment which zeroes out all the non-active charges
+   *
+   * There are multiple scenarios
+   *   - period between acquisition date and fulfilment date => None which indicates cancel now
+   *     - usually contractEffectiveDate to customerAcceptanceDate, except in the case of Guardian Weekly+6for6
+   *       where customerAcceptanceDate indicates start of GW proper invoiced period instead of start of 6for6
+   *       invoiced period despite GW+6for6 being just a regular Subscription with multiple products.
+   *     - free trial, or user choose start date of first issue in the future (lead time)
+   *   - Subscription within invoiced period proper => Some(endOfLastInvoicePeriod)
+   *   - free product => None which indicates cancel now
+   *   - edge case of being on the first day of invoice period however bill run has not yet happened => ERROR
+   *   - Today is after end of last invoice date and bill run has already completed => ERROR
```

Comment:
Can we have tests to prove these scenarios?

Reply:
Addressed by https://github.com/guardian/membership-common/pull/630